### PR TITLE
feat: add initial benchmarks

### PR DESCRIPTION
**Pull request by katcipis**
Initial results with these benchmarks:

```
 make bench
go test -bench=. -benchmem -memprofile=mem.prof -cpuprofile cpu.prof .
goos: linux
goarch: amd64
pkg: github.com/katcipis/jtoh
cpu: Intel(R) Core(TM) i5-8600K CPU @ 3.60GHz
BenchmarkNonJSONStreams/10_Messages-6              10000            313154 ns/op          326243 B/op       1276 allocs/op
BenchmarkNonJSONStreams/100_Messages-6               326           4408481 ns/op         3236302 B/op      12531 allocs/op
BenchmarkNonJSONStreams/1000_Messages-6               60          19985376 ns/op        32345204 B/op     125039 allocs/op
BenchmarkNonJSONStreams/10000_Messages-6               4         398242734 ns/op        323431336 B/op   1250051 allocs/op
BenchmarkJSONStreams/10_Messages-6                 25491             87333 ns/op           18896 B/op        194 allocs/op
BenchmarkJSONStreams/100_Messages-6                 2701            450165 ns/op          160074 B/op       1724 allocs/op
BenchmarkJSONStreams/1000_Messages-6                 163           6356340 ns/op         1571928 B/op      17026 allocs/op
BenchmarkJSONStreams/10000_Messages-6                 28          57814914 ns/op        15692935 B/op     170048 allocs/op
PASS
ok      github.com/katcipis/jtoh        17.205s
```

Amount of allocs/op seems O(N) with stream size for all scenarios, although that is not representative of an increase in overall memory usage, it just means new allocations, but if old objects are released and collected by the GC that is OK. It seems that Go benchmarks are limited to measuring the amount of allocs and total of memory allocated from all allocations, without taking into consideration the timeline and alloc/free relationships.

To prove this I built a much simpler code snippet with its own benchmark: https://github.com/katcipis/playground/tree/master/go/benchmarks

Running the benchmark for the dummy example:

```sh
make bench
go test -bench=. -benchmem
goos: linux
goarch: amd64
pkg: github.com/katcipis/playground/go/benchmarks
cpu: Intel(R) Core(TM) i5-8600K CPU @ 3.60GHz
BenchmarkMem/AllocCount10-6               305625              7681 ns/op           10240 B/op         10 allocs/op
BenchmarkMem/AllocCount100-6               13580             76656 ns/op          102400 B/op        100 allocs/op
BenchmarkMem/AllocCount1000-6               2113            798016 ns/op         1024002 B/op       1000 allocs/op
BenchmarkMem/AllocCount10000-6               142           8300732 ns/op        10240028 B/op      10000 allocs/op
PASS
ok      github.com/katcipis/playground/go/benchmarks    8.094s
```

Which is an increase linearly on allocs/op and also B/op. But if you check the code, there is no reference to the allocated memory after the test function finishes: 

https://github.com/katcipis/playground/blob/master/go/benchmarks/mem.go

```go
func Mem(count int) {
	var x []byte
	for i := 0; i < count; i++ {
		x = make([]byte, 1024)
		for i := range x {
			x[i] = 6
		}
	}
}
```

So each allocated array counts as an allocation op and accumulates, even though memory is not kept during the whole lifetime of the benchmark (GC should collect it at some point). Analyzing the memory profile created by this simple function is quite scary (just as it was with jtoh mem profile =P):

```
go tool pprof mem.prof
File: benchmarks.test
Type: alloc_space
Time: Jul 31, 2021 at 9:51pm (CEST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 9.91GB, 100% of 9.91GB total
Dropped 1 node (cum <= 0.05GB)
      flat  flat%   sum%        cum   cum%
    9.91GB   100%   100%     9.91GB   100%  github.com/katcipis/playground/go/benchmarks.Mem
         0     0%   100%     9.91GB   100%  github.com/katcipis/playground/go/benchmarks_test.BenchmarkMem.func1
         0     0%   100%     9.90GB 99.92%  testing.(*B).launch
         0     0%   100%     9.91GB   100%  testing.(*B).runN
```

It seems like the code is using a total of almost 10 GB of memory, but that is actually not true. Unless the GC doesn't work at all, this doesn't reflect memory usage in general, just a cumulative memory of all the allocations performed on the benchmark (even if the allocated memory was "released" right after it was allocated, apparently).

TL;DR: The results of the benchmark are scary at first, but they do not represent how much memory jtoh actually uses when processing streams.

Anyway, the amount of allocations also can be optimized, but that would be done on a future PR.

### CODE REVIEW

**Comment by katcipis** [comment]
just changed the scale of the tests and now I get this:

```
make bench
go test -bench=. -benchmem -memprofile=mem.prof -cpuprofile cpu.prof .
goos: linux
goarch: amd64
pkg: github.com/katcipis/jtoh
cpu: Intel(R) Core(TM) i5-8600K CPU @ 3.60GHz
BenchmarkNonJSONStreams/1000_Messages-6                      100          16140266 ns/op        32345400 B/op     125039 allocs/op
BenchmarkNonJSONStreams/10000_Messages-6                       5         352593951 ns/op        323434473 B/op   1250060 allocs/op
BenchmarkNonJSONStreams/100000_Messages-6                      1        2373150644 ns/op        3235005160 B/op 12500130 allocs/op
BenchmarkNonJSONStreams/1000000_Messages-6                     1        17747196160 ns/op       32367204456 B/op        125000310 allocs/op
BenchmarkJSONStreams/1000_Messages-6                         238           5965017 ns/op         1564000 B/op      17026 allocs/op
BenchmarkJSONStreams/10000_Messages-6                         16          69279376 ns/op        15611255 B/op     170053 allocs/op
BenchmarkJSONStreams/100000_Messages-6                         2         582491606 ns/op        156082152 B/op   1700287 allocs/op
BenchmarkJSONStreams/1000000_Messages-6                        1        7091232064 ns/op        1560950088 B/op 17002844 allocs/op
PASS
ok      github.com/katcipis/jtoh        37.503s
```

If I analyze on pprof:

```
make bench/mem/analyze
go tool pprof mem.prof
File: jtoh.test
Type: alloc_space
Time: Aug 6, 2021 at 6:38pm (CEST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 42.41GB, 99.46% of 42.64GB total
Dropped 28 nodes (cum <= 0.21GB)
Showing top 10 nodes out of 22
      flat  flat%   sum%        cum   cum%
   17.10GB 40.09% 40.09%    17.10GB 40.09%  encoding/json.(*Decoder).refill
   16.90GB 39.62% 79.71%    16.90GB 39.62%  github.com/katcipis/jtoh.(*bufferedReader).reset (inline)
    5.20GB 12.20% 91.91%     5.20GB 12.20%  encoding/json.NewDecoder (inline)
    1.31GB  3.07% 94.98%     1.54GB  3.62%  encoding/json.(*scanner).error (inline)
    1.16GB  2.71% 97.69%    42.64GB   100%  github.com/katcipis/jtoh.J.Do
    0.49GB  1.15% 98.85%     0.49GB  1.15%  reflect.mapassign
    0.22GB  0.52% 99.37%     0.23GB  0.55%  encoding/json.quoteChar
    0.04GB 0.087% 99.46%     0.63GB  1.47%  encoding/json.(*decodeState).object
         0     0% 99.46%     9.78GB 22.94%  encoding/json.(*Decoder).Decode
         0     0% 99.46%     9.49GB 22.24%  encoding/json.(*Decoder).More (inline)
```

Which accounts for 41GB of allocations, it is way more than the memory I have on my host (I don't have any swap...because fuck that). I also followed mem usage using top -p <pid> and it has a peak of 50MB (because of the current non-json parsing problem), and that is it, it falls down during execution to 10MB before exiting (releasing memory back to the OS). 

So it seems that the default benchmark output to use on pprof is not extremely suitable for detecting memory leaks...or maybe I'm missing something terribly basic x_x.

**Comment by katcipis** [comment]
Got something better than just measuring with top using **time**:

```
/usr/bin/time -v make bench
go test -bench=. -benchmem -memprofile=mem.prof -cpuprofile cpu.prof .
goos: linux
goarch: amd64
pkg: github.com/katcipis/jtoh
cpu: Intel(R) Core(TM) i5-8600K CPU @ 3.60GHz
BenchmarkNonJSONStreams/1000_Messages-6                      100          25285079 ns/op        32345639 B/op     125039 allocs/op
BenchmarkNonJSONStreams/10000_Messages-6                       3         384077366 ns/op        323431213 B/op   1250059 allocs/op
BenchmarkNonJSONStreams/100000_Messages-6                      1        2747941895 ns/op        3235016336 B/op 12500153 allocs/op
BenchmarkNonJSONStreams/1000000_Messages-6                     1        15021421162 ns/op       32367157960 B/op        125000276 allocs/op
BenchmarkJSONStreams/1000_Messages-6                         340           7193886 ns/op         1563990 B/op      17026 allocs/op
BenchmarkJSONStreams/10000_Messages-6                         16          67761186 ns/op        15611026 B/op     170051 allocs/op
BenchmarkJSONStreams/100000_Messages-6                         2         694657682 ns/op        156078236 B/op   1700288 allocs/op
BenchmarkJSONStreams/1000000_Messages-6                        1        6993580138 ns/op        1560775672 B/op 17002723 allocs/op
PASS
ok      github.com/katcipis/jtoh        35.806s
        Command being timed: "make bench"
        User time (seconds): 46.32
        System time (seconds): 2.23
        Percent of CPU this job got: 135%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:35.88
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 70852
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 2
        Minor (reclaiming a frame) page faults: 58403
        Voluntary context switches: 151739
        Involuntary context switches: 1298
        Swaps: 0
        File system inputs: 0
        File system outputs: 152
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

Memory usage really isn't that high (although we can bring it much lower for the non-json parsing scenario).

**Comment by i4ki** [comment]
Yeah, the allocations per op are just a hint to make programmers aware of how much memory is being allocated per run, because it affects performance (the whole point of bench). Not the best tool for looking for leaks.
Like maybe you implemented the best/faster algorithm but it is still slow because you're allocating too much, then `go bench` also shows the allocs/op there. About the pprof, it's the same for C or other languages, it just profile allocation numbers and sizes, but don't track them. 

indeed, it seems Go still lacks a good tool (or official tool) for leak detection. I found some from the community but they don't seem mature (I could be wrong, didn't test them).

Go supports MSAN (`go test -msan` works) but it'd be cool if they added support for ASAN because this way we would have a universal tool for memory errors that could track pointers crossing the boundaries C-Go. The clang ASAN (`-fsanitize=address`) works really well, is very easy to use, doesn't require any external tool, etc. I just found this proposal and active CLs to implement it in Go but still a work in progress: https://github.com/golang/go/issues/44853

**Comment by katcipis** [comment]
> Yeah, the allocations per op are just a hint to make programmers aware of how much memory is being allocated per run, because it affects performance (the whole point of bench). Not the best tool for looking for leaks.
> Like maybe you implemented the best/faster algorithm but it is still slow because you're allocating too much, then `go bench` also shows the allocs/op there. About the pprof, it's the same for C or other languages, it just profile allocation numbers and sizes, but don't track them.
> 
> indeed, it seems Go still lacks a good tool (or official tool) for leak detection. I found some from the community but they don't seem mature (I could be wrong, didn't test them).
> 
> Go supports MSAN (`go test -msan` works) but it'd be cool if they added support for ASAN because this way we would have a universal tool for memory errors that could track pointers crossing the boundaries C-Go. The clang ASAN (`-fsanitize=address`) works really well, is very easy to use, doesn't require any external tool, etc. I just found this proposal and active CLs to implement it in Go but still a work in progress: [golang/go#44853](https://github.com/golang/go/issues/44853)

yeah it does make sense to show total allocations made by the code, but since it is a memory profiler I imagined it would also provide something like Valgrind massif that can also show overall total memory/heap usage, not just the sum of all high level allocations. But in the end it is not that terrible, maybe my expectations were just off =P. The dream would be to be able to catch regressions on total memory usage and even catch leaks automatically, apparently doesn't feel feasible. Also kinda sad because the [runtime exports enough information](https://pkg.go.dev/runtime#MemProfileRecord) to implement something like that (that is why I imagined the mem profile would already have that data).
